stdlib/public/core/Zip.swift

```diff
@@ -136,13 +147,140 @@ extension Zip2Sequence.Iterator: IteratorProtocol {
+///
+/// - Note: Use `zip(_:_:)` if you don't require the latter elements of the
+/// longest sequence or if the length of both sequences are equal.
+///
+/// - Parameters:
+///   - sequence1: The first `Sequence`.
+///   - sequence2: The second `Sequence`.
+///
+/// - Returns: An instance of `ZipLongest2Sequence`.
+@inlinable // FIXME(sil-serialize-all)
+public func zipLongest<Sequence1, Sequence2>(
+  _ sequence1: Sequence1, _ sequence2: Sequence2
+) -> ZipLongest2Sequence<Sequence1, Sequence2> {
+  return ZipLongest2Sequence(_sequence1: sequence1, _sequence2: sequence2)
+}
+
+/// A sequence of pairs created from a pair of sequences by zipping their
+/// elements until the *longest* sequence is exhausted, padding the latter
+/// elements of the shortest sequence with `nil`.
+///
+/// - Note: Instances of `ZipLongest2Sequence` can be created with the global
+/// `zipLongest(_:_:)` function.
+@_fixed_layout // FIXME(sil-serialize-all)
+public struct ZipLongest2Sequence<Sequence1 : Sequence, Sequence2 : Sequence> {
+  @usableFromInline // FIXME(sil-serialize-all)
+  internal let _sequence1: Sequence1
+  @usableFromInline // FIXME(sil-serialize-all)
+  internal let _sequence2: Sequence2
+
+  /// Constructs an instance of `ZipLongest2Sequence`.
+  @inlinable // FIXME(sil-serialize-all)
+  public // @testable
+  init(_sequence1 sequence1: Sequence1, _sequence2 sequence2: Sequence2) {
+    (_sequence1, _sequence2) = (sequence1, sequence2)
+  }
+}
+
+extension ZipLongest2Sequence {
+  /// The iterator for `ZipLongest2Sequence`.
+  ///
+  /// - Note: To create an instance, use `makeIterator()`.
+  @_fixed_layout // FIXME(sil-serialize-all)
+  public struct Iterator {
+    @usableFromInline // FIXME(sil-serialize-all)
+    internal var _baseStream1: Sequence1.Iterator
+    @usableFromInline // FIXME(sil-serialize-all)
+    internal var _baseStream2: Sequence2.Iterator
+    @usableFromInline // FIXME(sil-serialize-all)
+    internal var _reachedEnd: Bool = false
+
+    /// Constructs an instance of `ZipLongest2Sequence.Iterator`.
+    @inlinable // FIXME(sil-serialize-all)
+    internal init(
+      _ iterator1: Sequence1.Iterator,
+      _ iterator2: Sequence2.Iterator
+    ) {
+      (_baseStream1, _baseStream2) = (iterator1, iterator2)
+    }
+  }
+}
+
+extension ZipLongest2Sequence.Iterator: IteratorProtocol {
+  /// The type of element returned by `next()`.
+  public typealias Element = (Sequence1.Element?, Sequence2.Element?)
+
+  /// Advances to the next element and returns it, or `nil` if no next element
+  /// exists. Once `nil` has been returned, all subsequent calls return `nil`.
+  @inlinable // FIXME(sil-serialize-all)
+  public mutating func next() -> Element? {
+    if _reachedEnd {
+      return nil
+    }
+
+    let element1 = _baseStream1.next()
+    let element2 = _baseStream2.next()
+
+    if element1 == nil && element2 == nil {
+      _reachedEnd = true
+      return nil
+    }
+
+    return (element1, element2)
```

Comment:
Using `} else {` for the return is probably cleaner here.

Reply:
`return element1 == nil && element2 == nil ? nil : (element1, element2)` is also an option.